src/chainstate/stacks/auth.rs

```diff
@@ -21,44 +21,1431 @@ use net::StacksMessageCodec;
+    /// In doing so, recover all public keys and verify that they hash to the signer
+    /// via the given hash mode.
+    /// Returns the final sighash
+    pub fn verify(&self, initial_sighash: &Txid, cond_code: &TransactionAuthFlags) -> Result<Txid, net_error> {
+        let (pubkey, next_sighash) = TransactionSpendingCondition::next_verification(initial_sighash, cond_code, &self.key_encoding, &self.signature)?;
+        let addr_bytes = match StacksAddress::from_public_keys(0, &self.hash_mode.to_address_hash_mode(), 1, &vec![pubkey]) {
+            Some(a) => {
+                a.bytes
+            }
+            None => {
+                return Err(net_error::VerifyingError("Failed to generate address from public key".to_string()));
+            }
+        };
+        
+        if addr_bytes != self.signer {
+            return Err(net_error::VerifyingError(format!("Public key hash {} does not match signer hash {}", &addr_bytes.to_hex(), &self.signer.to_hex())));
+        }
+
+        Ok(next_sighash)
+    }
+}
+
+impl StacksMessageCodec for TransactionSpendingCondition {
+    fn serialize(&self) -> Vec<u8> {
+        match *self {
+            TransactionSpendingCondition::Singlesig(ref data) => data.serialize(),
+            TransactionSpendingCondition::Multisig(ref data) => data.serialize()
+        }
+    }
+
+    fn deserialize(buf: &Vec<u8>, index_ptr: &mut u32, max_size: u32) -> Result<TransactionSpendingCondition, net_error> {
+        let mut index = *index_ptr;
+
+        if (buf.len() as u32) <= index {
+            return Err(net_error::UnderflowError);
+        }
+
+        // NOTE: this takes advantage of the fact that the first byte of each type variant's
+        // serialized byte representation -- the hash mode -- uniquely identifies the variant.
+        let hash_mode_u8 = buf[index as usize];
+        let cond = match hash_mode_u8 {
+            x if x == SinglesigHashMode::P2PKH as u8 || x == SinglesigHashMode::P2WPKH as u8 => {
+                let cond = SinglesigSpendingCondition::deserialize(buf, &mut index, max_size)?;
+                TransactionSpendingCondition::Singlesig(cond)
+            }
+            x if x == MultisigHashMode::P2SH as u8 || x == MultisigHashMode::P2WSH as u8 => {
+                let cond = MultisigSpendingCondition::deserialize(buf, &mut index, max_size)?;
+                TransactionSpendingCondition::Multisig(cond)
+            }
+            _ => {
+                test_debug!("Invalid hash mode {}", hash_mode_u8);
+                return Err(net_error::DeserializeError)
+            }
+        };
+
+        *index_ptr = index;
+        Ok(cond)
+    }
+}
+
+impl TransactionSpendingCondition {
+    pub fn new_singlesig_p2pkh(pubkey: StacksPublicKey) -> Option<TransactionSpendingCondition> {
+        let signer_addr = match StacksAddress::from_public_keys(0, &AddressHashMode::SerializeP2PKH, 1, &vec![pubkey.clone()]) {
```

Comment:
The `?` operator also works with `Option<..>` types and can be pretty helpful in lines like this

Reply:
Didn't know that; will update.

Reply:
Addressed in d399b6d9113660aa1a71e88d9e7915c6dd8bb12f